pkg/querier/querier.go

```diff
@@ -68,6 +68,13 @@ type Config struct {
 	UseSecondStoreBeforeTime flagext.Time `yaml:"use_second_store_before_time"`
 }
 
+// UserError are errors caused by user input as opposed to errors executing the query.
+type UserError string
```

Comment:
I think it's great to add a specific type here, but most similar errors are currently using `httpgrpc.Errorf()`.
https://github.com/cortexproject/cortex/blob/5ede3635eb550e57ed22a8d0d48c8fe783036b54/pkg/querier/queryrange/query_range.go#L34-L36

Therefore I think you should either do the same here, or replace all similar ones with your new type.  
Perhaps it should go in package `validation`?

Reply:
> but most similar errors are currently using httpgrpc.Errorf()

I think we use `httpgrpc.Errorf()` when directly returning from a gRPC call (or it's parent is a direct gRPC call), but in this case it's a function called by the PromQL engine, which then wraps the error again, so it would look weird to me having a `httpgrpc.Errorf()` wrapped by a PromQL error.

I personally think that for this fix, having a dedicated error type is just fine and we could consider refactoring existing errors as a separate work. However, I would also argue that this error may be better placed in the `validation` package and maybe renamed to `LimitError` (`UserError` looks very generic to me).

Reply:
> I think we use httpgrpc.Errorf() when directly returning from a gRPC call (or it's parent is a direct gRPC call), but in this case it's a function called by the PromQL engine, which then wraps the error again, so it would look weird to me having a httpgrpc.Errorf() wrapped by a PromQL error.

I was about to phrase that similarly as well - `NewQueryRange` is used by the engine directly.


> However, I would also argue that this error may be better placed in the validation package and maybe renamed to `LimitError` 

I think this is sensible.

Reply:
`httpgrpc.Errorf()` is used in non-gRPC situations, e.g. in `Distributor.Push()`.
To be clear, I hate httpgrpc, but I also hate inconsistency.

Reply:
I merged this PR because I didn't notice this comment, sorry! Let me open an issue.

Reply:
https://github.com/cortexproject/cortex/issues/3181

Reply:
Apologies @bboreham, think I've missed that case - I'll take a deeper look and follow up on this.

Reply:
I don't understand what point you are making in "can't control the status code".
You added a case to the type switch not far above the line where it already decodes httpgrpc errors:
https://github.com/cortexproject/cortex/blob/f1ac40e05e9e9d8d4d16c61121b3a91848d4ec56/pkg/api/queryable.go#L42-L49